tripal/t4d8
# Automatic Creation of Single Valued, String Chado Fields

## Issues Addressed
Issue tripal/t4d8#129
Issue tripal/t4d8#236 
Issue tripal/t4d8#237

Wait to review this until after the merge of PR tripal/t4d8#254 because this PR has that code in it and all of the changes in that PR are also showing up here.

## Description
This PR provides new functionality for field creation. It will automatically add all single-valued string fields to all content types during the Chado Prepare step.   The only exception is the 'Pub' content type. That one requires some special handling.  You can also create new entities for content types where the string fields are sufficient to meet the unique constraint (e.g. Organism).

This PR does not implement the `ChadoStorage::deleteValues()` function (issue tripal/tripal#1350) because there is not yet a distinction between an "Unpublish" and a "Delete" so I thought it would be best to not implement a delete until we had the buttons.  The  `ChadoStorage::findValues()` function is also not yet implemented (issue tripal/tripal#1351).

This PR also does not support integer fields, text fields or other database fields other than `character varying`.  I left a "@todo" comment in the code to finish those up.  But this PR is provided now so that everyone can have access to the new code. 

## How to Test
1. Uninstall Tripal, pull this branch and reinstall Tripal (or start from a brand new installation)
2. Install Chado
3. Prepare Chado. It should indicate all of the content types are created.  It's a bit slower creating content types now because it takes time to create all the fields.  
4. Go to Admin > Structure > Tripal Content Types. Pick any content type.  You will see the set of fields that have been added.  They should not appear in the 'Disabled' section but should be in the Content section.
5. Go to Content > Tripal Content > Add new Tripal Content and pick the Organism content type. Test creation and editing of the content type. 
6. Look in the Chado database in the organism table. A new record should exist in the organism table, and any changes you made should be there.
7. Return to the Admin > Structure > Tripal Content Types. You should be able to move the fields around on the manage display and manage form.  
8. Check out the field settings and the field storage settings. You should see the controlled vocabulary term that each field is assigned and the Chado table/column that the field maps to.
